### DISCUSSION TITLE
What does the controller.cmd_thrust variable actually mean?

### DISCUSSION THREAD

**Discussion by maxwell-lucas2** [bitcraze/discussions#1471] [Q&A]
I am running experiments with the Crazyflie 2.0 and am consistently running into an issue where once resistance is met the quadcopter automatically flips over in the roll direction.

I have been logging controller.cmd_thrust and some other variables to check if it is a saturation issue and see that the value automatically hits its limit of 65535 a few seconds before loosing control but I need to nail down exactly where I can find how this variable is used and computed in the code.

If the number is maxed out does that mean that one or more motors are at full throttle? I'm newly exploring quadcopter controller so any help or more detail on this variable would be greatly appreciated!

![Screenshot 2024-07-16 195541](https://github.com/user-attachments/assets/c05104fb-d696-4531-a5ab-0230e7d5e5a4)

Here is a quick screen shot of the crazyflie maxing out this value for reference.

**Comment by evoggy** [comment]
Hi!

Is it just roll, not pitch behaving like this?

There's more information about the stabilizer [here](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/) if you haven't had a chance to look at it yet. Are you using the default PID controller?

I quickly looked though the code and from what I understand this is the commanded thrust, but it will be capped to the value you're seeing. The links below are for the PID:

Here's where it's being used to do the calculation for the motor output: https://github.com/bitcraze/crazyflie-firmware/blob/9ee56072ec1a769ab6dac3e1a1a781ef105853ba/src/modules/src/power_distribution_quadrotor.c#L92

And here's where it's being capped: https://github.com/bitcraze/crazyflie-firmware/blob/9ee56072ec1a769ab6dac3e1a1a781ef105853ba/src/modules/src/power_distribution_quadrotor.c#L150

**Reply by maxwell-lucas2** [reply]
Thank you so much for the reply,

Yes I am using the default PID controller. That is very helpful, would you happen to have any idea where the roll, pitch, and yaw values are specifically calculated at each time step before they are mixed in power_distribution_quadrotor.c?

Thank you for the help!!

**Reply by knmcguire** [reply]
That should be right here in the PID controller: https://github.com/bitcraze/crazyflie-firmware/blob/41e8f3c2004a872d0ffea166818a1530df07f787/src/modules/src/controller/controller_pid.c#L127-L129
